core/domain/auth_jobs_one_off.py

```diff
@@ -39,6 +39,258 @@
+            return
+        # Else: key is a non-empty auth_id.
+
+        auth_id = key
+
+        reports = list(cls.report_assoc_collisions(auth_id, assoc_info_pairs))
+        if reports:
+            for report in reports:
+                yield (cls.ERROR_ASSOC_INCONSISTENCIES, report)
+            return
+        # Else: 1 <= len(assoc_info_pairs) <= 3.
+
+        assoc_info = dict(assoc_info_pairs)
+
+        reports = list(cls.report_assoc_inconsistencies(auth_id, **assoc_info))
+        if reports:
+            for report in reports:
+                yield (cls.ERROR_ASSOC_INCONSISTENCIES, report)
+            return
+        # Else: assoc_by_auth_id and assoc_by_user_id are either:
+        #   1. Both deleted.
+        #   2. Both present and contain consistent values.
+
+        # NOTE: This is an arbitrary choice, could use ASSOC_BY_USER_ID_KEY too.
+        assoc_key = cls.ASSOC_BY_AUTH_ID_KEY
+
+        user_is_permanently_deleted, firebase_account_exists = (
+            assoc_key not in assoc_info,
+            cls.FIREBASE_ACCOUNT_DISABLED_KEY in assoc_info)
+
+        if user_is_permanently_deleted:
+            if firebase_account_exists:
+                firebase_admin.auth.update_user(auth_id, disabled=True)
+                firebase_admin.auth.delete_user(auth_id)
+            return
+        # Else: assoc_by_auth_id and assoc_by_user_id are both present and have
+        # consistent values.
+
+        user_id, marked_as_deleted = assoc_info[assoc_key]
+
+        if not firebase_account_exists:
+            user_settings = user_models.UserSettingsModel.get(user_id)
+            firebase_admin.auth.create_user(
+                uid=auth_id, email=user_settings.email,
+                # NOTE: Even though the user might be marked for deletion, it's
+                # important to create a disabled Firebase account anyway so that
+                # the same email can not be claimed while the deletion request
+                # is pending.
+                disabled=marked_as_deleted)
+        elif marked_as_deleted != assoc_info[cls.FIREBASE_ACCOUNT_DISABLED_KEY]:
+            firebase_admin.auth.update_user(auth_id, disabled=marked_as_deleted)
+
+    @classmethod
+    def report_assocs_missing(cls, assoc_info_pairs):
+        """Yields debug information for each of the given associations.
+
+        NOTE: Since assoc_by_auth_id and Firebase accounts are keyed by auth_id,
+        the only kinds of associations that could ever hit this function are
+        UserAuthDetailsModel.
+
+        Args:
+            assoc_info_pairs: list(tuple(str, (str, bool))). The list of
+                associations that do not correspond to any auth_id.
+
+        Yields:
+            str. A debug string for each association.
+        """
+        for _, (user_id, _) in assoc_info_pairs:
+            yield (
+                'UserAuthDetailsModel(id="%s") does not correspond to a unique '
```

Comment:
Should "a unique" be changed to "any"?

Reply:
I don't think so, I used this wording to stay consistent with the analogous "UserIdByFirebaseAuthIdModel does not correspond to a unique" message, and because "a unique" already implies "any".

Reply:
Ah, just so you know, the impression I got from reading this was that it corresponded to 2+ items.

Reply:
Right, that's why they don't correspond to "a unique" ID; they _do_ have an ID assigned, though.

Should I change anything to clear this up?

Reply:
I'm confused, sorry. Doesn't report_assocs_missing() only get called when the user ID doesn't correspond to a firebase_auth_id at all? Hence why I thought it should have been "any".

Reply:
(I mean, the interpretation of 2+ firebase auth IDs is wrong, right?)

Reply:
Ah right, I got confused about which line we were talking about.

In the end I just removed the unique qualifier and added a signal that `firebase_auth_id=None`, since I don't feel too strongly about it.